// tailwindcss/src/dist/constants.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultPostCssConfigStubFile = exports.simpleConfigStubFile = exports.defaultConfigStubFile = exports.supportedPostCssConfigFile = exports.supportedConfigFiles = exports.cjsPostCssConfigFile = exports.cjsConfigFile = exports.defaultPostCssConfigFile = exports.defaultConfigFile = exports.cli = void 0;

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var cli = 'tailwind';
exports.cli = cli;
var defaultConfigFile = './tailwind.config.js';
exports.defaultConfigFile = defaultConfigFile;
var defaultPostCssConfigFile = './postcss.config.js';
exports.defaultPostCssConfigFile = defaultPostCssConfigFile;
var cjsConfigFile = './tailwind.config.cjs';
exports.cjsConfigFile = cjsConfigFile;
var cjsPostCssConfigFile = './postcss.config.cjs';
exports.cjsPostCssConfigFile = cjsPostCssConfigFile;
var supportedConfigFiles = [cjsConfigFile, defaultConfigFile];
exports.supportedConfigFiles = supportedConfigFiles;
var supportedPostCssConfigFile = [cjsPostCssConfigFile, defaultPostCssConfigFile];
exports.supportedPostCssConfigFile = supportedPostCssConfigFile;

var defaultConfigStubFile = _path["default"].resolve(__dirname, '../stubs/defaultConfig.stub.js');

exports.defaultConfigStubFile = defaultConfigStubFile;

var simpleConfigStubFile = _path["default"].resolve(__dirname, '../stubs/simpleConfig.stub.js');

exports.simpleConfigStubFile = simpleConfigStubFile;

var defaultPostCssConfigStubFile = _path["default"].resolve(__dirname, '../stubs/defaultPostCssConfig.stub.js');

exports.defaultPostCssConfigStubFile = defaultPostCssConfigStubFile;
