// yaml/browser/dist/dist/legacy-exports.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "binary", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.b;
  }
});
Object.defineProperty(exports, "floatTime", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.f;
  }
});
Object.defineProperty(exports, "intTime", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.i;
  }
});
Object.defineProperty(exports, "omap", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.o;
  }
});
Object.defineProperty(exports, "pairs", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.p;
  }
});
Object.defineProperty(exports, "set", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.s;
  }
});
Object.defineProperty(exports, "timestamp", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.t;
  }
});
Object.defineProperty(exports, "warnFileDeprecation", {
  enumerable: true,
  get: function get() {
    return _warningsDf54cb.c;
  }
});

var _warningsDf54cb = require("./warnings-df54cb69.js");

require("./PlainValue-b8036b75.js");

require("./resolveSeq-492ab440.js");
