// yaml/browser/dist/dist/types.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Alias", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.A;
  }
});
Object.defineProperty(exports, "Collection", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.C;
  }
});
Object.defineProperty(exports, "Merge", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.M;
  }
});
Object.defineProperty(exports, "Node", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.N;
  }
});
Object.defineProperty(exports, "Pair", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.P;
  }
});
Object.defineProperty(exports, "Scalar", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.S;
  }
});
Object.defineProperty(exports, "YAMLMap", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.d;
  }
});
Object.defineProperty(exports, "YAMLSeq", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.Y;
  }
});
Object.defineProperty(exports, "binaryOptions", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.b;
  }
});
Object.defineProperty(exports, "boolOptions", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.a;
  }
});
Object.defineProperty(exports, "intOptions", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.i;
  }
});
Object.defineProperty(exports, "nullOptions", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.n;
  }
});
Object.defineProperty(exports, "strOptions", {
  enumerable: true,
  get: function get() {
    return _resolveSeq492ab.s;
  }
});
Object.defineProperty(exports, "Schema", {
  enumerable: true,
  get: function get() {
    return _SchemaE94716c.S;
  }
});

var _resolveSeq492ab = require("./resolveSeq-492ab440.js");

var _SchemaE94716c = require("./Schema-e94716c8.js");

require("./PlainValue-b8036b75.js");

require("./warnings-df54cb69.js");
